Use go modules, analysis package and add test

bombsimon opened the pull request:
* Use go modules
* Use `analysis.Analyzer` for the linter
* Introduce running mode to use from `golangci-lint`
* Add tests using the testing package and assert fixer
* Add standalone linter (`cmd/whitespace`)

---

Sorry for the big PR and rewrite! While doing some refactoring I renamed `main.go` to `whitespace.go` (since it's not a `main` package) but it seems like `git` doesn't realize it's a rename as well so the whole file is diffing. As you can see most of the original code is intact and the changes are described below.

Given the current requirements in `golangci-lint` and per the discussion in https://github.com/golangci/golangci-lint/issues/3967, linters needs to use the `analysis` package. This PR mostly just adapts to that but to also support fixes from the analyzer itself (which only makes sense since we at one point want to address https://github.com/golangci/golangci-lint/issues/1779). To make the analysis fixer work I did some additional changes:

* There's now a difference between where the diagnostic/error will be marked (`Message.Diagnostic`), where a fix will start which will be the `LBrace` _or_ following comment (`Message.FixStart`) and where the first statement which is where the fix will end (`Message.FixEnd`)
* To avoid having `golangci-lint` to use the cache to read the file and find the line number, I added `Message.LineNumber` which tells which line number should be deleted when removing whitespaces
* `MessageType` is now slimmed down to only being either `MessageTypeRemove` or `MessageTypeAdd`

To allow the run method to return diagnostics in a way `golangci-lint` want I added support for a running mode which will be set to `RunningModeGolangCI` when called from `golangci-lint`.

There's actually no need for this change to add the `cmd/whitespace` binary but given how simple it is to make an executable that also supports fixing I did so anyway.

Another benefit of using the `analysis` package is the ease of testing. Given that I added tests for different configuration to ensure the linter does what it's supposed to do.

I already prepared a PR in `golangci-lint` which will make use of the changes in this PR: https://github.com/golangci/golangci-lint/pull/4003

bombsimon committed:
Use go modules, analysis package and add test

* Use go modules
* Use `analysis.Analyzer` for the linter
* Introduce running mode to use from `golangci-lint`
* Add tests using the testing package and assert fixer
* Add standalone linter (`cmd/whitespace`)

ldez commented:
It can be a good idea to add GitHub action.

bombsimon committed:
Fix comments and typos, add another test

bombsimon committed:
Add GitHub actions

robinknaapen commented on whitespace.go:
```diff
@@ -0,0 +1,283 @@
+			v.messages = append(v.messages, Message{
+				Diagnostic:  stmt.Lbrace,
+				FixStart:    stmt.List[0].Pos(),
+				FixEnd:      stmt.List[0].Pos(),
```
It this correct? Or should this be `.End()`?

bombsimon replied:
Yeah it's a bit weird maybe but why I added the split between where the diagnostic is and where the fix is. Basically this just says that from where the first statement starts, add a newline. So it's not replacing anything because the range is zero, it's just adding.

```go
func fn() {
    stmt.List[0]
   ^
   |__From here
   |__ To here (same place)
    Add a "\n"
}
```

This will result in

```go
func fn() {
    "\n"stmt.List[0]
}
```

Which is

```go
func fn() {
    
stmt.List[0]
}
```

And since applying suggestions with the `-fix` flag also formats the code according to `go fmt`, the result will be

```go
func fn() {
    
	stmt.List[0]
}
```

There should be test cases for all the fixes in the `testdata` directory, including one that adds a newline like this one. The `.golden` file shows what's expected.

robinknaapen replied:
Great explanation, thanks for taking the time

robinknaapen commented:
I also found a "issue" in whitespace

When not using gofmt. It's possible to create multiple whitespaces

```golang
package main

import "fmt"

func fn3() {
    // A comment
                   // < Whitespace will target this line
                   // < This is not seen by whitespace, thus fix will only omit the previous line, resulting in yet another linter warning?
    fmt.Println(`Hello World`)
}
```

I don't know if this is important enough to fix. Because majority of people use gofmt...

bombsimon commented:
> I don't know if this is important enough to fix. Because majority of people use gofmt...

This is a great question! And I don't think I can answer this. I discussed it briefly in the `golangci-lint` PR here: https://github.com/golangci/golangci-lint/pull/4003.

Also as seen in my my comment https://github.com/ultraware/whitespace/pull/8#discussion_r1307829107, applying the suggested fixes with the `-fix` flag will _also_ reformat the file according to `go fmt` so if you want to keep your non formatted file it's not possible to use the analyzer. Not sure if that's a blocker for you.

Here's an example package that I formatted, showing other changes as well:

Before with inconsistent spaces:

```text
› bat -A example.go
───────┬─────────────────────────────────────
       │ File: example.go
───────┼─────────────────────────────────────
   1   │ package·x␊
   2   │ ␊
   3   │ import·"fmt"␊
   4   │ ␊
   5   │ func·fn()·{␊
   6   │ ├──┤····␊
   7   │ ····fmt.Println("Hello,·World")␊
   8   │ ␊
   9   │ ········fmt.Println("Hello,·World")␊
  10   │ }␊
───────┴─────────────────────────────────────
```

And after applying suggested fixes:
```text
› bat -A example.go
───────┬─────────────────────────────────
       │ File: example.go
───────┼─────────────────────────────────
   1   │ package·x␊
   2   │ ␊
   3   │ import·"fmt"␊
   4   │ ␊
   5   │ func·fn()·{␊
   6   │ ├──┤fmt.Println("Hello,·World")␊
   7   │ ␊
   8   │ ├──┤fmt.Println("Hello,·World")␊
   9   │ }␊
───────┴─────────────────────────────────
```

Personally I have a hard time seeing the target audience for someone that's interested in a formatter like this but don't care for `go fmt` but I have no data to back that up, just a gut feeling.

To avoid false positives I could rewrite this to actually get the position of where the comment ends instead of the `LBrace` but otherwise just running the linter twice should fix the problem.

robinknaapen commented:
@bombsimon 

> Also as seen in my my comment https://github.com/ultraware/whitespace/pull/8#discussion_r1307829107, applying the suggested fixes with the -fix flag will also reformat the file according to go fmt

Thats good information, I wasn't aware of the fact that the `-fix` flag also fixes like gofmt
Is this done before or after analysis. 

If it's before, than the issue will is non-existent


> To avoid false positives I could rewrite this to actually get the position of where the comment ends instead of the LBrace but otherwise just running the linter twice should fix the problem.

This is "beaking" in a logical sense. But I don't think it's an impact at all. I'll ask my colleague for his opinion to this

@elwint, what is your thought on this?

bombsimon commented:
> Is this done before or after analysis.

Sadly it looks like after.

Before:

```sh
› bat -A example.go
───────┬───────────────────────────────────────
       │ File: example.go
───────┼───────────────────────────────────────
   1   │ package·main␊
   2   │ ␊
   3   │ ␊
   4   │ ␊
   5   │ ␊
   6   │ import·"fmt"␊
   7   │ ␊
   8   │ func·main()·{␊
   9   │ ␊
  10   │ ␊
  11   │ ␊
  12   │ ····fmt.Println("Hello,·World")␊
  13   │ ␊
  14   │ ····├──┤fmt.Println("Hello,·World·2")␊
  15   │ }␊
───────┴───────────────────────────────────────
```

After (only one run)

```sh
› bat -A example.go
───────┬───────────────────────────────────
       │ File: example.go
───────┼───────────────────────────────────
   1   │ package·main␊
   2   │ ␊
   3   │ import·"fmt"␊
   4   │ ␊
   5   │ func·main()·{␊
   6   │ ␊
   7   │ ├──┤fmt.Println("Hello,·World")␊
   8   │ ␊
   9   │ ├──┤fmt.Println("Hello,·World·2")␊
  10   │ }␊
───────┴───────────────────────────────────
```

This is actually an issue for `golangci-lint` as well since I only store _one_ line to remove. Maybe it would actually be petter to store the exact position of the ennd of the last node or comment and the start of the first node or commennt to have the exact range. We could also store a list of lines to remove for `golangci-lint`. However, the comment linked above (https://github.com/golangci/golangci-lint/pull/4003) will still be an issue although more related to `golangci-lint` than this linter. I think if we had `pos.End()` of the last comment or node we could use that one to add a newline and also solve the issue in `golangci-lint` so that's two reasons to actually rewrite this.

robinknaapen commented:
@bombsimon 

First of all, thanks for being patient.

I think that if it's doable than why not at this point

bombsimon commented:
> First of all, thanks for being patient.

Thank _you_ for taking the time to review this and considering the PR 😃 

> I think that if it's doable than why not at this point

Sure thing, I'll give this a go when I get the time!

bombsimon committed:
Make linter work on non formatted files

Instead of relying on the file being `gofmt`:ed and only containing one
empty line at most this commit will keep track of comments after the
opening bracket of block statements and set the fix start to where this
comment ends. This means that we will fix all the way from the left
bracket or comment til the first statement no matter how many empty
lines this is.

This will also keep a list of lines affected by the linter to support
fixing multiple empty lines in `golangci-lint`.

bombsimon commented:
Wow, can't believe this took me almost a month, sorry about that!

Anyway, I added a change so we now keep track of where the opening position is; either the left bracket or the end of a comment on the same line. This made me able to set the whole fix range between where this node ends and where the first statement (or comment) starts.

While doing this I also changed `LineNumber` to `LineNumbers` so we can fix the whole range in `golangci-lint` as well and not only a single line.

I added some test cases where the file is not formatted and contains multiple lines. Sadly since the `analysistest` package requires the `// want` directive I cannot write a unit test (easily) that checks this works without comments but I just made a simple test file locally:

```go
package main

import "fmt"

func main() {


    fmt.Println("Hello, World")


}
```

And ran `go run ./cmd/whitespace -fix ./testpkg` and it formatted the file correctly.

robinknaapen commented on testdata/src/whitespace_no_multiline/no_multiline.go:
```diff
@@ -48,3 +48,49 @@ func fn2(
+
+
+
+        fmt.Println("No cmoments")
```
Small typo

robinknaapen commented on whitespace.go:
```diff
@@ -41,10 +41,10 @@ type Message struct {
 
-	// LineNumber represent the actual line number in the file. This is set when
-	// finding the diagnostic to make it easier to suggest fixes in
+	// LineNumbers represent the actual line number in the file. This is set
```
```diff
- // LineNumbers represent the actual line number in the file. This is set
+ // LineNumbers represent the actual line numbers in the file. This is set
  // when finding the diagnostic to make it easier to suggest fixes in
  // golangci-lint.
```

robinknaapen commented on testdata/src/whitespace_no_multiline/no_multiline.go:
```diff
@@ -48,3 +48,49 @@ func fn2(
+
+
+        
+} // want "unnecessary trailing newline"
```
I've added 2 tests 

```golang
// Regular func (FuncDecl) that's not `gofmt`:ed. with comments
func fn5() { 
 	// A comment that should still exist after this
	// This one should also still exist 
		  
	  

	fmt.Println("Hello, World")

	if true { 
		// A comment that should still exist after this
		// This one should also still exist



		fmt.Println("No cmoments")


	} // want "unnecessary trailing newline"



		
} // want "unnecessary trailing newline"

// Regular func (FuncDecl) that's not `gofmt`:ed. with comment blocks
func fn6() { 
	/*
		Lorem ipsum dolor sit amet, consectetur adipiscing elit.
		Curabitur ornare dolor at nulla ultrices cursus.
		Mauris pharetra metus ac condimentum sodales.
		Fusce viverra libero vitae tellus dictum, sed congue risus sodales.
	*/
		  
	  

	fmt.Println("Hello, World")

	if true {
		/* 
			Lorem ipsum dolor sit amet, consectetur adipiscing elit.
			Curabitur ornare dolor at nulla ultrices cursus.
			Mauris pharetra metus ac condimentum sodales.
			Fusce viverra libero vitae tellus dictum, sed congue risus sodales.
		*/ 


		fmt.Println("No cmoments")


	} // want "unnecessary trailing newline"



		
} // want "unnecessary trailing newline"
```

The thing is that the file is correctly formatted after running Whitespace but I am missing diagnostic warnings?

When I ran Whitespace, I go this output: (omitted out of context diagnostics)

``` 
/home/robin/dev/github.com/ultraware/whitespace/testdata/src/whitespace_no_multiline/no_multiline.go:121:1: unnecessary trailing newline
/home/robin/dev/github.com/ultraware/whitespace/testdata/src/whitespace_no_multiline/no_multiline.go:116:2: unnecessary trailing newline
/home/robin/dev/github.com/ultraware/whitespace/testdata/src/whitespace_no_multiline/no_multiline.go:153:1: unnecessary trailing newline
/home/robin/dev/github.com/ultraware/whitespace/testdata/src/whitespace_no_multiline/no_multiline.go:148:2: unnecessary trailing newline
```

This does not match the trend set by previous tests. For example [fn4](https://github.com/ultraware/whitespace/pull/8/commits/d9dd7ed2b510ff9409a7b39a40a9a98cca917160#diff-df158a7786b3b6e51a5ed562f2d8df485305e3ecf5db73bf0d5831ee7beb40a3R84) shows a proper diagnostic warning

```
/home/robin/dev/github.com/ultraware/whitespace/testdata/src/whitespace_no_multiline/no_multiline.go:84:49: unnecessary leading newline
```


Please correct me if I'm wrong there :sweat:

bombsimon replied:
These changes shouldn't really be touched by `whitespace` since they're not violating `whitespace` rules. A comment is seen as any statement and there are no whitespaces between the opening block and the first comment. However since the file isn't properly `gofmt`:ed the fixer will format it and replace all occurrences with more than one empty line with only one empty line. Because of this `whitespace` will not report any "Unnecessary leading newline" because there is no whitespace between the opening of the block (`{`) and the first statement or comment. I don't think `whitespace` should warn about newlines in this case right?

I added your tests with proper `want` directives (just like you did them) and the test passes and like you said formats the file correctly.

Regarding `fn4` (and I guess all test cases); I interpreted it as by design to _not_ see comments on the same line as a statement since that will still result in an empty line to the eye. This means that this is OK:

```go
func fn1() {
    // Comment is on new line

    fmt.Println("ok")
}

func fn2() {
    /*
        Multiline
    */

    fmt.Println("ok")
}
```

But this is not:

```go
func fn1() { // Comment is on same line as `LBrace`

    fmt.Println("not ok")
}

func fn2() { /* Multiline */

    fmt.Println("not ok")
}
```

This is interesting though, what do you think about this case?
```go
func fn1() /* Multiline
    comment */

    fmt.Println("ok?")
}
```

Am I missing something here regarding either the intended way `whitespace` should work or your question? Do you think the tests I added in my last commit is what you expected?

robinknaapen replied:
Ah, I see where I went wrong. I thought golangci was responsible for fixing. But it's analysis. Thanks for clearing that up

> This is interesting though, what do you think about this case?

That's a great question. First off, I have never seen that kind of comment in the wild

I assume the linter triggers in this case.

Personally I think this should not trigger a warning. But I'll have to gather some thoughs since this is undefined. I'll ask my colleagues asap.

What's your opinion @bombsimon

bombsimon replied:
> I thought golangci was responsible for fixing. But it's analysis. Thanks for clearing that up

Well, you're not wrong. Currently `golangci-lint` can't interpret the reports so it's both. When we run `whitespace` as a standalone linter with `-fix` it's the analysis package, when we run `golangci-lint` with `--fix` it's some custom code. That's the reason why I'm keeping which lines. But I don't think that's related, bottom line is that if it's one or more empty lines between the bracket line and the first comment or statement, both will report (and fix). If not, no matter how many lines are _after_ that initial comment or statement `whitespace` won't add a diagnostic.

> I assume the linter triggers in this case.

Yes, currently it does. But I think it's just a side effect and not intentional. In fact, I think the current check for comments is wrong, this line should check if the comment _starts_ on the same line, not if it ends.

I think I agree. Since this is OK:

```go
func Fn() { // Never use Fn with flux capacitor
    // Lorem ipsum dolor sit amet

    fmt.Println("")
}
```

I don't see why this should not be:

```go
func Fn() { /* Never use Fn with flux capacitor
    Lorem ipsum dolor sit amet */

    fmt.Println("")
}
```

robinknaapen commented:
@bombsimon 

> Wow, can't believe this took me almost a month, sorry about that!

Don't mention it, I really appreciate your expertise. Thanks for all the effort :muscle:

bombsimon committed:
Fix typos, add tests from PR

robinknaapen commented:
@bombsimon 

Sorry for the delay.

> Yes, currently it does. But I think it's just a side effect and not intentional. In fact, I think the current check for comments is wrong, this line should check if the comment starts on the same line, not if it ends.

The consensus about this case is that it should indeed be allowed.

--- 

After this I think that's all

bombsimon committed:
Multiline comments ending after opening pos is ok

robinknaapen commented:
@bombsimon 

Nice solution :muscle: 

First thing in the morning I will review and ask a colleague for review.

Thanks for all the effort once again

bombsimon commented:
> Sorry for the delay.

No worries, there's no rush here and I appreciate you're considering this PR.

> Nice solution 💪
> 
> First thing in the morning I will review and ask a colleague for review.

Thanks! Yeah the change grew over time and I understand that this is a lot to review. But I hope that all the test cases should help add some confidence that the code is actually doing what it's supposed to do. If you can think of any more test cases that would probe the expected behaviour feel free to add them!

> Thanks for all the effort once again

Thank's for the great feedback and considering this PR! Just let me know if there's anything else missing or that needs clarification.

bombsimon committed:
Fix typo

ldez commented:
Can you create a tag? (I suggest v0.1.0)

robinknaapen commented:
@ldez 

Yes, I will. Thanks for reminding :sweat_smile:
